components/TextFields/src/MDCTextField.m

```diff
@@ -0,0 +1,530 @@
+
+@property(nonatomic, strong) UIImage *clearButtonImage;
+@property(nonatomic, strong) MDCTextInputLayoutCoordinator *coordinator;
+@property(nonatomic, readonly, weak) UIButton *internalClearButton;
+
+@end
+
+@implementation MDCTextField
+
+@dynamic borderStyle;
+
+@synthesize internalClearButton = _internalClearButton;
+
+- (instancetype)initWithFrame:(CGRect)frame {
+  self = [super initWithFrame:frame];
+  if (self) {
+    _coordinator = [[MDCTextInputLayoutCoordinator alloc] initWithTextInput:self];
+
+    [self commonMDCTextFieldInitialization];
+  }
+  return self;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)aDecoder {
+  self = [super initWithCoder:aDecoder];
+  if (self) {
+    NSString *interfaceBuilderPlaceholder = super.placeholder;
+    [self commonMDCTextFieldInitialization];
+
+    _clearButtonImage = [aDecoder decodeObjectForKey:MDCTextFieldClearButtonImageKey];
+    _clearButtonColor = [aDecoder decodeObjectForKey:MDCTextFieldClearButtonColorKey];
+    _coordinator = [aDecoder decodeObjectForKey:MDCTextFieldCoordinatorKey]
+                       ?: [[MDCTextInputLayoutCoordinator alloc] initWithTextInput:self];
+
+    if (interfaceBuilderPlaceholder.length) {
+      self.placeholder = interfaceBuilderPlaceholder;
+    }
+    [self setNeedsLayout];
+  }
+  return self;
+}
+
+- (instancetype)initWithLeftView:(UIView *)leftView {
+  self = [self initWithFrame:CGRectZero];
+  if (self) {
+    self.leftView = leftView;
+  }
+  return self;
+}
+
+- (void)dealloc {
+  NSNotificationCenter *defaultCenter = [NSNotificationCenter defaultCenter];
+  [defaultCenter removeObserver:self name:UITextFieldTextDidBeginEditingNotification object:self];
+  [defaultCenter removeObserver:self name:UITextFieldTextDidChangeNotification object:self];
+}
+
+- (void)encodeWithCoder:(NSCoder *)aCoder {
+  [super encodeWithCoder:aCoder];
+  [aCoder encodeObject:self.clearButtonColor forKey:MDCTextFieldClearButtonColorKey];
+  [aCoder encodeObject:self.clearButtonImage forKey:MDCTextFieldClearButtonImageKey];
+  [aCoder encodeConditionalObject:self.coordinator forKey:MDCTextFieldCoordinatorKey];
+}
+
+- (instancetype)copyWithZone:(NSZone *)zone {
+  MDCTextField *copy = [[[self class] alloc] init];
+
+  copy.clearButtonColor = self.clearButtonColor.copy;
+  copy.clearButtonImage = self.clearButtonImage.copy;
+
+  // Just a pointer value copies.
+  copy.coordinator = self.coordinator;
+
+  return copy;
+}
+
+- (void)commonMDCTextFieldInitialization {
+
+  [super setBorderStyle:UITextBorderStyleNone];
+
+  // Set the clear button color to black with 54% opacity.
+  [self setClearButtonColor:[UIColor colorWithWhite:0 alpha:[MDCTypography captionFontOpacity]]];
+
+  NSNotificationCenter *defaultCenter = [NSNotificationCenter defaultCenter];
+  [defaultCenter addObserver:self
+                    selector:@selector(textFieldDidBeginEditing:)
+                        name:UITextFieldTextDidBeginEditingNotification
+                      object:self];
```

Comment:
Yeah. Unfortunately, if I don't add self as the object, it will hear every single `UITextFieldTextDidBeginEditingNotification` which means I'd have to put introspection into `textFieldDidBeginEditing:` to see if the one that began editing is self. This does that for free.